test/jdk/java/nio/channels/DatagramChannel/SRTest.java

```diff
@@ -24,184 +24,216 @@
+                throw new RuntimeException("ClassicWriter threw exception: " + e);
+            } finally {
+                log.println("ClassicWriter finished");
             }
         }
+
+        @Override
+        public void close() throws IOException {
+            ds.close();
+        }
     }
 
-    public static class NioWriter implements Sprintable {
+    public static class NioWriter implements Runnable, AutoCloseable {
+        final DatagramChannel dc;
         final int port;
-        Exception e = null;
 
-        NioWriter(int port) {
+        NioWriter(int port) throws IOException {
+            this.dc = DatagramChannel.open();
             this.port = port;
         }
 
-        public void throwException() throws Exception {
-            if (e != null)
-                throw e;
-        }
-
         public void run() {
             try {
-                DatagramChannel dc = DatagramChannel.open();
                 ByteBuffer bb = ByteBuffer.allocateDirect(256);
-                bb.put("hello".getBytes());
+                bb.put(DATA_STRING.getBytes(US_ASCII));
                 bb.flip();
-                InetAddress address = InetAddress.getLocalHost();
+                InetAddress address = InetAddress.getLoopbackAddress();
                 InetSocketAddress isa = new InetSocketAddress(address, port);
                 dc.send(bb, isa);
-                Thread.sleep(50);
-                dc.send(bb, isa);
             } catch (Exception ex) {
-                e = ex;
+                log.println("NioWriter [" + dc.socket().getLocalAddress() + "]");
+                throw new RuntimeException("NioWriter threw exception: " + ex);
+            } finally {
+                log.println("NioWriter finished");
             }
         }
+
+        @Override
+        public void close() throws IOException {
+            dc.close();
+        }
     }
 
-    public static class ClassicReader implements Sprintable {
+    public static class ClassicReader implements Runnable, AutoCloseable {
         final DatagramSocket ds;
-        Exception e = null;
 
         ClassicReader() throws IOException {
-            this.ds = new DatagramSocket();
+            InetSocketAddress address = new InetSocketAddress(InetAddress.getLoopbackAddress(), 0);
+            this.ds = new DatagramSocket(address);
         }
 
         int port() {
             return ds.getLocalPort();
         }
 
-        public void throwException() throws Exception {
-            if (e != null)
-                throw e;
-        }
-
         public void run() {
             try {
                 byte[] buf = new byte[256];
                 DatagramPacket dp = new DatagramPacket(buf, buf.length);
                 ds.receive(dp);
-                String received = new String(dp.getData());
-                log.println(received);
-                ds.close();
+                String received = new String(dp.getData(), 0, DATA_STRING.length(), US_ASCII);
```

Comment:
That's not quite correct either. The `DatagramPacket` should have its length and offset positioned correctly after `receive` - that's the value you should use (`p.getOffset()`, `p.getLength()`).

Reply:
Ah ok understood, I can see that is more correct now. Will change shortly